File: algorithm/oc/next-permutation.m
// 31. Next Permutation
// Difficulty: Medium
// Implement next permutation, which rearranges numbers into the lexicographically next greater permutation of numbers.

// If such arrangement is not possible, it must rearrange it as the lowest possible order (ie, sorted in ascending order).

// The replacement must be in-place, do not allocate extra memory.

// Here are some examples. Inputs are in the left-hand column and its corresponding outputs are in the right-hand column.
// 1,2,3 → 1,3,2
// 3,2,1 → 1,2,3
// 1,1,5 → 1,5,1

// Time:  O(n)
// Space: O(1)

#import <Foundation/Foundation.h>

void nextPermutation(NSMutableArray* nums) {
  if ([nums count] < 2) {
    return;
  }
  
  NSInteger i;
  // Find the first element (pivot) which is less than its successor.
  for (i = [nums count] - 2; i >= 0; i--) {
    if ([nums[i] integerValue] < [nums[i + 1] integerValue]) {
      break;
    }
  }
  
  if (i >= 0) {
    // Find the number which is first greater than pivot from reverse start, and swap it with pivot
    for (NSInteger j = [nums count] - 1; j >= i; j--) {
      if (nums[j] > nums[i]) {
        [nums exchangeObjectAtIndex:j withObjectAtIndex:i];
        break;
      }
    }
  }
  // Make the sequence after pivot non-descending
  [nums replaceObjectsInRange:NSMakeRange(i + 1, nums.length - i - 1)
         withObjectsFromArray:[[[nums subarrayWithRange:NSMakeRange(i + 1, nums.length - i - 1)] reverseObjectEnumerator] allObject]];
}

File: zZ/java/next-permutation.java
/**
 * Next Permutation
 *  
 * Implement next permutation, which rearranges numbers into the lexicographically next greater permutation of numbers.
 * 
 * If such arrangement is not possible, it must rearrange it as the lowest possible order (ie, sorted in ascending order).
 * 
 * The replacement must be in-place, do not allocate extra memory.
 * 
 * Here are some examples. Inputs are in the left-hand column and its corresponding outputs are in the right-hand column.
 * 1,2,3 → 1,3,2
 * 3,2,1 → 1,2,3
 * 1,1,5 → 1,5,1
 */

public class Solution {
    public void nextPermutation(int[] nums) {
        if (nums == null) {
            throw new NullPointerException();
        }

        if (nums.length == 0 || nums.length == 1) {
            return;
        }

        int i = nums.length - 1;
        while (i - 1 >= 0 && nums[i - 1] >= nums[i]) {
            --i;
        }

        if (i == 0) {
            for (int j = 0; j < nums.length / 2; ++j) {
                swap(nums, j, nums.length - 1 - j);
            }
        } else {
            int j = nums.length - 1;
            for (; nums[j] <= nums[i - 1]; --j);
            swap(nums, i - 1, j);
            Arrays.sort(nums, i, nums.length);
        }
    }

    private void swap(int[] nums, int i, int j) {
        nums[i] = nums[i] ^ nums[j];
        nums[j] = nums[i] ^ nums[j];
        nums[i] = nums[i] ^ nums[j];
    }
}